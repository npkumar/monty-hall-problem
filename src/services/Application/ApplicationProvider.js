import React, { createContext, useReducer } from "react";
import PropTypes from "prop-types";

// Actions
const SET_SIMULATIONS = "SET_SIMULATIONS";
const SET_IS_CHANGE_DOOR = "SET_IS_CHANGE_DOOR";
const SET_WIN_PERCENTAGE = "SET_WIN_PERCENTAGE";
const SET_IS_SIMULATION_RUNNING = "SET_IS_SIMULATION_RUNNING";

export const ApplicationContext = createContext(null);

function reducer(state, action) {
  switch (action.type) {
    case SET_SIMULATIONS: {
      return { ...state, simulations: action.payload, winPercentage: null };
    }
    case SET_IS_CHANGE_DOOR: {
      return { ...state, isChangeDoor: action.payload, winPercentage: null };
    }
    case SET_WIN_PERCENTAGE: {
      return { ...state, winPercentage: action.payload };
    }
    case SET_IS_SIMULATION_RUNNING: {
      return { ...state, isSimulationRunning: action.payload };
    }
    default:
      throw new Error();
  }
}
const ApplicationProvider = ({ children }) => {
  const initialState = {
    simulations: 4000,
    isChangeDoor: true,
    winPercentage: null,
    isSimulationRunning: false,
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const value = {
    application: state,
    setSimulations: (simulations) =>
      dispatch({ type: SET_SIMULATIONS, payload: simulations }),
    setIsChangeDoor: (isChangeDoor) =>
      dispatch({ type: SET_IS_CHANGE_DOOR, payload: isChangeDoor }),
    setWinPercentage: (winPercentage) =>
      dispatch({ type: SET_WIN_PERCENTAGE, payload: winPercentage }),
    setIsSimulationRunning: (isSimulationRunning) =>
      dispatch({
        type: SET_IS_SIMULATION_RUNNING,
        payload: isSimulationRunning,
      }),
  };

  return (
    <ApplicationContext.Provider value={value}>
      {children}
    </ApplicationContext.Provider>
  );
};

ApplicationProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default ApplicationProvider;
