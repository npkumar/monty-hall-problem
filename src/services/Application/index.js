// Hooks
export { default as useApplication } from "./useApplication";

// Providers
export { default as ApplicationProvider } from "./ApplicationProvider";
