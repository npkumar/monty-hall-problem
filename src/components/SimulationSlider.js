import React from "react";
import { withStyles, makeStyles } from "@material-ui/core/styles";
import Slider from "@material-ui/core/Slider";
import Typography from "@material-ui/core/Typography";
import { useApplication } from "../services/Application";

const useStyles = makeStyles((theme) => ({
  root: {
    width: "auto",
    margin: "0 1rem",
  },
  banner: {
    fontSize: 48,
    textAlign: "center",
    fontWeight: "bold",
    marginBottom: 36,
  },
  label: {
    fontSize: 32,
    display: "inline-flex",
    border: "8px solid #52af77",
    textAlign: "center",
    minHeight: "10rem",
    minWidth: "10rem",
    borderRadius: "100%",
    justifyContent: "center",
    alignItems: "center",
    flexDirection: "column",
  },
  labelText: {
    marginTop: 8,
  },
  subtitle: {
    fontSize: 12,
    fontWeight: "normal",
  },
}));

const PrettoSlider = withStyles({
  root: {
    color: "#52af77",
    height: 8,
  },
  thumb: {
    height: 24,
    width: 24,
    backgroundColor: "#fff",
    border: "2px solid currentColor",
    marginTop: -8,
    marginLeft: -12,
    "&:focus, &:hover, &$active": {
      boxShadow: "inherit",
    },
    "&.MuiSlider-thumb": {
      marginTop: -8,
      marginLeft: -12,
      height: 24,
      width: 24,
    },
  },
  active: {},
  valueLabel: {
    left: "calc(-50% + 4px)",
  },
  track: {
    height: 8,
    borderRadius: 4,
  },
  rail: {
    height: 8,
    borderRadius: 4,
  },
})(Slider);

export default function SimulationSlider() {
  const classes = useStyles();
  const {
    application: { isSimulationRunning, simulations },
    setSimulations,
  } = useApplication();

  const handleSliderChange = (event, newValue) => {
    setSimulations(newValue);
  };

  return (
    <div className={classes.root}>
      <Typography component="div" className={classes.banner}>
        <div className={classes.label}>
          <div className={classes.labelText} data-testid="simulations">
            {simulations}
          </div>
          <div className={classes.subtitle}>#simulations</div>
        </div>
      </Typography>
      <PrettoSlider
        data-testid="slider"
        valueLabelDisplay="off"
        aria-label="pretto slider"
        value={simulations}
        onChange={handleSliderChange}
        min={1}
        max={10000000}
        disabled={isSimulationRunning}
      />
    </div>
  );
}
