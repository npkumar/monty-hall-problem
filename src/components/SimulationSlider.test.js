import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import SimulationSlider from "./SimulationSlider";
import { ApplicationContext } from "../services/Application/ApplicationProvider";

const value = {
  application: { isSimulationRunning: false, simulations: 4000 },
};

describe("SimulationSlider", () => {
  afterEach(() => cleanup());

  it("renders", () => {
    const { asFragment } = render(
      <ApplicationContext.Provider value={value}>
        <SimulationSlider />
      </ApplicationContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should have simulations and slider", () => {
    const { getByTestId } = render(
      <ApplicationContext.Provider value={value}>
        <SimulationSlider />
      </ApplicationContext.Provider>
    );
    expect(getByTestId("simulations")).toBeInTheDocument();
    expect(getByTestId("slider")).toBeInTheDocument();
    expect(getByTestId("simulations")).toHaveTextContent(4000);
  });
});
