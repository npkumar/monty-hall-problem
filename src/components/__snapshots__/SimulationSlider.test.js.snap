// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimulationSlider renders 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-1"
  >
    <div
      class="MuiTypography-root makeStyles-banner-2 MuiTypography-body1"
    >
      <div
        class="makeStyles-label-3"
      >
        <div
          class="makeStyles-labelText-4"
          data-testid="simulations"
        >
          4000
        </div>
        <div
          class="makeStyles-subtitle-5"
        >
          #simulations
        </div>
      </div>
    </div>
    <span
      class="MuiSlider-root WithStyles(ForwardRef(Slider))-root-6 MuiSlider-colorPrimary"
      data-testid="slider"
    >
      <span
        class="MuiSlider-rail WithStyles(ForwardRef(Slider))-rail-11"
      />
      <span
        class="MuiSlider-track WithStyles(ForwardRef(Slider))-track-10"
        style="left: 0%; width: 0.0399900039990004%;"
      />
      <input
        type="hidden"
        value="4000"
      />
      <span
        aria-label="pretto slider"
        aria-orientation="horizontal"
        aria-valuemax="10000000"
        aria-valuemin="1"
        aria-valuenow="4000"
        class="MuiSlider-thumb WithStyles(ForwardRef(Slider))-thumb-7 MuiSlider-thumbColorPrimary"
        data-index="0"
        role="slider"
        style="left: 0.0399900039990004%;"
        tabindex="0"
      />
    </span>
  </div>
</DocumentFragment>
`;
